### PR TITLE
feat(hmac-auth): Add validate request body for hmac auth plugin

### PR DESCRIPTION
**Pull request by arthur-zhang**
### What this PR does / why we need it:

Add validate request body for hmac auth plugin, by calculate body hmac hash and put it in Digest Header 

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/plugins/hmac-auth.lua [hunk 1]
```diff
@@ -193,6 +206,19 @@ local function do_nothing(v)
     return v
 end
 
+local function validate_body(ctx, secret_key, params, req_body)
+    if not req_body then
+        req_body = ""
+    end
+    local digest_header = core.request.header(ctx, DIGEST)
```
Comment (starsz) [inline apisix/plugins/hmac-auth.lua:213]: `digest` header looks not good for me.

File: docs/zh/latest/plugins/hmac-auth.md [hunk 2]
```diff
@@ -272,6 +284,29 @@ Accept-Ranges: bytes
...
+     -d "{\"hello\":\"world\"}"
+
+HTTP/1.1 200 OK
+Content-Type: text/html; charset=utf-8
+Transfer-Encoding: chunked
+Connection: keep-alive
+Date: Tue, 19 Jan 2021 11:33:20 GMT
+Server: APISIX/2.2
+......
+```
```
Comment (starsz) [inline docs/zh/latest/plugins/hmac-auth.md:308]: Can you help to update the English doc?

Comment (arthur-zhang) [reply docs/zh/latest/plugins/hmac-auth.md:308]: I will

File: apisix/plugins/hmac-auth.lua (hunk 1 above)
Comment (arthur-zhang) [reply apisix/plugins/hmac-auth.lua:213]: The reason I use this header is that:
The Digest response HTTP header provides a digest of the requested resource.

File: apisix/plugins/hmac-auth.lua [hunk 3]
```diff
@@ -193,6 +206,19 @@ local function do_nothing(v)
     return v
 end
 
+local function validate_body(ctx, secret_key, params, req_body)
+    if not req_body then
+        req_body = ""
+    end
```
Comment (tzssangglass) [inline apisix/plugins/hmac-auth.lua:212]:
```suggestion
    req_body = req_body or ""
```

File: apisix/plugins/hmac-auth.lua [hunk 4]
```diff
@@ -193,6 +206,19 @@ local function do_nothing(v)
...
+    local digest_header = core.request.header(ctx, DIGEST)
+    if not digest_header then
+        -- it's ok if there is no digest header and no body
+        return req_body == ""
+    end
```
Comment (tzssangglass) [inline apisix/plugins/hmac-auth.lua:217]: Can we judge this first?

File: docs/zh/latest/plugins/hmac-auth.md [hunk 5]
```diff
@@ -186,6 +188,16 @@ print(base64.b64encode(hash.digest()))
 | --------- | -------------------------------------------- |
 | SIGNATURE | 8XV1GB7Tq23OJcoz6wjqTs4ZLxr9DiLoY4PxzScWGYg= |
 
+### Body 校验
+
+把 `validate_request_body` 设置为 true 来进行请求 body 的校验。 插件将计算 hmac-sha 值，对比头部中的 Digest 头部值。
```
Comment (tzssangglass) [inline docs/zh/latest/plugins/hmac-auth.md:193]:
```suggestion
把 `validate_request_body` 设置为 true 来进行请求 body 的校验。插件将计算 hmac-sha 值，对比头部中的 Digest 头部值。
```

Comment (tzssangglass) [reply docs/zh/latest/plugins/hmac-auth.md:193]: `Digest` in code is `digest`, better to keep same.

File: apisix/plugins/hmac-auth.lua (hunk 4 above)
Comment (arthur-zhang) [reply apisix/plugins/hmac-auth.lua:215]: Maybe not, because we can omit digest header when body is empty.

File: docs/zh/latest/plugins/hmac-auth.md (hunk 5 above)
Comment (arthur-zhang) [reply docs/zh/latest/plugins/hmac-auth.md]: done

File: apisix/plugins/hmac-auth.lua [hunk 6]
```diff
@@ -193,6 +206,19 @@ local function do_nothing(v)
     return v
 end
 
+local function validate_body(ctx, secret_key, params, req_body)
+    if not req_body then
+        req_body = ""
+    end
+    local digest_header = core.request.header(ctx, DIGEST)
+    if not digest_header then
```
[the remark this answers is not in the document]
Comment (arthur-zhang) [reply apisix/plugins/hmac-auth.lua:212]: Maybe not, because we can omit digest header when body is empty.

**Comment by tzssangglass** [comment]
hi @arthur-zhang, I notice this `当无请求 body 时，可不传 Digest 头部，网关会校验是否确实无请求 body。如果要传 Digest 头部，可计算长度为 0 的空字符串的 hmac-sha 值。`, should we add test cases about GET method?

**Comment by arthur-zhang** [comment]
> hi @arthur-zhang, I notice this `当无请求 body 时，可不传 Digest 头部，网关会校验是否确实无请求 body。如果要传 Digest 头部，可计算长度为 0 的空字符串的 hmac-sha 值。`, should we add test cases about GET method?

I Add test `=== TEST 4: missing digest header and body is empty` for post quest without body, GET is no difference.

File: apisix/plugins/hmac-auth.lua [hunk 7]
```diff
@@ -79,7 +82,17 @@ local consumer_schema = {
...
             default = true,
-        }
+        },
+        validate_request_body = {
+            type = "boolean",
+            title = "A boolean value telling the plugin to enable body validation.",
+            default = false,
+        },
+        max_req_body = {
+            type = "number",
```
Comment (spacewander) [inline apisix/plugins/hmac-auth.lua:92]:
```suggestion
            type = "integer",
```

File: docs/en/latest/plugins/hmac-auth.md [hunk 8]
```diff
@@ -51,6 +51,8 @@ The `consumer` then adds its key to request header to verify its request.
...
 | keep_headers      | boolean       | optional    | false         | [ true, false ]                             | Whether it is necessary to keep the request headers of `X-HMAC-SIGNATURE`, `X-HMAC-ALGORITHM` and `X-HMAC-SIGNED-HEADERS` in the http request after successful authentication. true: means to keep the http request header, false: means to remove the http request header.   |
 | encode_uri_params | boolean       | optional    | true          | [ true, false ]                             | Whether to encode the uri parameter in the signature, for example: `params1=hello%2Cworld` is encoded, `params2=hello,world` is not encoded. true: means to encode the uri parameter in the signature, false: not to encode the uri parameter in the signature.               |
+| validate_request_body | boolean  | optional   | false         | [ true, false ]                             | Whether to check request body. |
+| max_req_body     | number        | optional   | 512KB         |                                             | Max allowed body size. |
```
Comment (spacewander) [inline docs/en/latest/plugins/hmac-auth.md:55]: Better to use a number instead of string in the doc's default value cell

File: docs/en/latest/plugins/hmac-auth.md [hunk 9]
```diff
@@ -278,6 +290,29 @@ Accept-Ranges: bytes
...
+     -H 'X-HMAC-SIGNATURE: ivlwjZPoVdSVvdSSM4drEFk9q9HS2jeJ5cAN9JffmdA=' \
+     -H 'X-HMAC-ALGORITHM: hmac-sha256' \
+     -H 'Date: Tue, 24 Aug 2021 03:19:21 GMT' \
+     -H 'X-HMAC-SIGNED-HEADERS: User-Agent;Digest' \
+     -H 'User-Agent: curl/7.29.0' \
+     -H 'Digest: L9b/+QMvhvnoUlSw5vq+kHPqnZiHGl61T8oavMVTaC4=' \
+     -H 'Content-Type: text/plain; charset=utf-8' \
+     -d "{\"hello\":\"world\"}"
+
+HTTP/1.1 200 OK
```
Comment (spacewander) [inline docs/en/latest/plugins/hmac-auth.md:307]: Better to only show the first line? APISIX 2.2 doesn't support this feature. We can avoid confusing users by only show the first line.

File: apisix/plugins/hmac-auth.lua [hunk 10]
```diff
@@ -29,13 +29,16 @@ local hmac       = require("resty.hmac")
 local consumer   = require("apisix.consumer")
 local plugin     = require("apisix.plugin")
 local ngx_decode_base64 = ngx.decode_base64
+local ngx_encode_base64 = ngx.encode_base64
 
+local DIGEST = "Digest"
```
Comment (spacewander) [inline apisix/plugins/hmac-auth.lua:34]: Can we make this header configurable and default to "X-HMAC-DIGEST"?

There is already a Digest header in the HTTP standard and it is not relative to the hmac.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Digest

File: docs/en/latest/plugins/hmac-auth.md (hunk 9 above)
Comment (arthur-zhang) [reply docs/en/latest/plugins/hmac-auth.md:307]: Got it

File: apisix/plugins/hmac-auth.lua (hunk 10 above)
Comment (arthur-zhang) [reply apisix/plugins/hmac-auth.lua:34]: OK

**Review by tzssangglass — APPROVED** [review]

File: apisix/plugins/hmac-auth.lua [hunk 11]
```diff
@@ -327,6 +351,18 @@ local function validate(ctx, params)
         return nil, {message = "Invalid signature"}
     end
 
+    local validate_request_body = get_conf_field(params.access_key, "validate_request_body")
+    if validate_request_body then
+        local max_req_body = get_conf_field(params.access_key, "max_req_body")
```
Comment (nic-chen) [inline apisix/plugins/hmac-auth.lua:356]: I think we could return here if no params.body_digest.

File: t/plugin/hmac-auth3.t [hunk 12]
```diff
@@ -521,3 +520,134 @@ passed
...
+
+            local signature = hmac:new(secret_key, hmac.ALGOS.SHA256):final(signing_string)
+            local body_digest = hmac:new(secret_key, hmac.ALGOS.SHA256):final(body)
+
+            core.log.info("signature:", ngx_encode_base64(signature))
+            local headers = {}
+            headers["X-HMAC-SIGNATURE"] = ngx_encode_base64(signature)
+            headers["X-HMAC-ALGORITHM"] = "hmac-sha256"
+            headers["Date"] = gmt
+            headers["X-Digest"] = ngx_encode_base64(body_digest)
```
Comment (spacewander) [inline t/plugin/hmac-auth3.t:569]: Should be `X-HMAC-DIGEST`?

File: docs/en/latest/plugins/hmac-auth.md [hunk 13]
```diff
@@ -192,6 +194,16 @@ print(base64.b64encode(hash.digest()))
 | --------- | -------------------------------------------- |
 | SIGNATURE | 8XV1GB7Tq23OJcoz6wjqTs4ZLxr9DiLoY4PxzScWGYg= |
 
+### Request body checking
+
+When `validate_request_body` is assigned to `true`, the plugin will check the request body. The plugin will calculate the hmac-sha value of the request body，and check against the `X-HMAC-DIGEST` header.
```
Comment (spacewander) [inline docs/en/latest/plugins/hmac-auth.md:199]: Better to mention the max_req_body limitation

File: docs/en/latest/plugins/hmac-auth.md [hunk 14]
```diff
@@ -192,6 +194,16 @@ print(base64.b64encode(hash.digest()))
...
+
+```
+X-HMAC-DIGEST: base64(hmac-sha(<body>))
+```
+
+When there is no request body, the `X-HMAC-DIGEST` header can be omitted. If you want to send request with this header whether the body is empty or not, you can set `X-HMAC-DIGEST` value to the hmac-sha of empty string.
```
Comment (spacewander) [inline docs/en/latest/plugins/hmac-auth.md:205]:
```suggestion
When there is no request body, the `X-HMAC-DIGEST` header can be omitted. If you want to send a request with this header when the body is missing, you can set `X-HMAC-DIGEST` value to the hmac-sha of empty string.
```

File: t/plugin/hmac-auth3.t (hunk 12 above)
Comment (arthur-zhang) [reply t/plugin/hmac-auth3.t:569]: Any string here is ok, I will change it to X-HMAC-DIGEST.

File: docs/en/latest/plugins/hmac-auth.md (hunk 14 above)
Comment (arthur-zhang) [reply docs/en/latest/plugins/hmac-auth.md]: done

File: docs/en/latest/plugins/hmac-auth.md (hunk 13 above)
Comment (arthur-zhang) [reply docs/en/latest/plugins/hmac-auth.md:199]: I add a **note** to that

File: apisix/plugins/hmac-auth.lua (hunk 11 above)
Comment (arthur-zhang) [reply apisix/plugins/hmac-auth.lua:356]: I think we should check whether the  params.body_digest exists or not.

Comment (nic-chen) [reply apisix/plugins/hmac-auth.lua:356]: when validate_request_body == true and no params.body_digest, why not return false?

Comment (arthur-zhang) [reply apisix/plugins/hmac-auth.lua:356]: When there is no request body, the `X-HMAC-DIGEST` header can be omitted.

File: apisix/plugins/hmac-auth.lua [hunk 15]
```diff
@@ -79,7 +82,17 @@ local consumer_schema = {
...
-        }
+        },
+        validate_request_body = {
+            type = "boolean",
+            title = "A boolean value telling the plugin to enable body validation.",
+            default = false,
+        },
+        max_req_body = {
+            type = "integer",
+            title = "Max request body allowed",
```
Comment (tokers) [inline apisix/plugins/hmac-auth.lua:93]: "Max request body size" would be better.

Comment (arthur-zhang) [reply apisix/plugins/hmac-auth.lua]: OK

File: apisix/plugins/hmac-auth.lua (hunk 11 above)
Comment (nic-chen) [reply apisix/plugins/hmac-auth.lua:356]: @arthur-zhang 

OK. Thanks for your explanation. How about when `validate_request_body` is enabled, `X-APISIX-HMAC-BODY-DIGEST` is required? Just pass empty strings for it when no body.

File: docs/en/latest/plugins/hmac-auth.md [hunk 16]
```diff
@@ -192,6 +194,18 @@ print(base64.b64encode(hash.digest()))
...
+
+When `validate_request_body` is assigned to `true`, the plugin will check the request body. The plugin will calculate the hmac-sha value of the request body，and check against the `X-HMAC-DIGEST` header.
+
+```
+X-HMAC-DIGEST: base64(hmac-sha(<body>))
+```
+
+When there is no request body, the `X-HMAC-DIGEST` header can be omitted. If you want to send a request with this header when the body is missing, you can set `X-HMAC-DIGEST` value to the hmac-sha of empty string.
+
+**Note:** The plugin will load the request body to memory to calculate the digest of the request body, which wight cause high memory consumption with large bodies. You can limit the max allowed body size by the configuration of `max_req_body`(default=512KB), request body larger than that will be rejected.
```
Comment (spacewander) [inline docs/en/latest/plugins/hmac-auth.md:207]:
```suggestion
**Note:** The plugin will load the request body to memory to calculate the digest of the request body, which might cause high memory consumption with large bodies. You can limit the max allowed body size by the configuration of `max_req_body`(default=512KB), request body larger than that will be rejected.
```

File: apisix/plugins/hmac-auth.lua (hunk 11 above)
Comment (arthur-zhang) [reply apisix/plugins/hmac-auth.lua:349]: Fix it

**Review by spacewander — APPROVED** [review]

**Review by nic-chen — APPROVED** [review]

**Review by tokers — APPROVED** [review]

**Review by starsz — APPROVED** [review]
